src/interface/term.ml

```diff
@@ -1007,3 +1007,155 @@ module type Smtlib_Bitv = sig
+
+  val fp_neg : t -> t
+  (** negation *)
+
+  val fp_add : t -> t -> t -> t
+  (** [fp_add rm f1 f2] addition *)
+
+  val fp_sub : t -> t -> t -> t
+  (** [fp_sub rm f1 f2] subtraction *)
+
+  val fp_mul : t -> t -> t -> t
+  (** [fp_mul rm f1 f2] multiplication *)
+
+  val fp_div : t -> t -> t -> t
+  (** [fp_mul rm f1 f2] division *)
+
+  val fp_fma : t -> t -> t -> t -> t
+  (** [fp_mul rm f1 f2] fused multiplication and addition *)
+
+  val fp_sqrt : t -> t -> t
+  (** [fp_sqrt rm f] square root *)
+
+  val fp_rem : t -> t -> t
+  (** [fp_rem f1 f2] remainder *)
+
+  val fp_roundToIntegral : t -> t -> t
+  (** [fp_roundToIntegral rm f] rounding to integral *)
+
+  val fp_min : t -> t -> t
+  (** [fp_min f1 f2] minimum *)
+
+  val fp_max : t -> t -> t
+  (** [fp_max f1 f2] maximum *)
+
+  val fp_leq : t -> t -> t
+  (** [fp_leq f1 f2] less or equal floating point comparison *)
+
+  val fp_lt : t -> t -> t
+  (** [fp_lt f1 f2] less than floating point comparison *)
+
+  val fp_geq : t -> t -> t
+  (** [fp_geq f1 f2] greater or equal floating point comparison *)
+
+  val fp_gt : t -> t -> t
+  (** [fp_lt f1 f2] greater than floating point comparison *)
+
+  val fp_eq : t -> t -> t
+  (** [fp_eq f1 f2] floating point equality *)
+
+  val fp_isNormal : t -> t
+  (** [fp_isNormal f] test if it is a normal floating point *)
+
+  val fp_isSubnormal : t -> t
+  (** [fp_isSubnormal f] test if it is a subnormal floating point *)
+
+  val fp_isZero : t -> t
+  (** [fp_isZero f] test if it is a zero *)
+
+  val fp_isInfinite : t -> t
+  (** [fp_isInfinite f] test if it is an infinite *)
+
+  val fp_isNaN : t -> t
+  (** [fp_isNaN f] test if it is NaN *)
+
+  val fp_isNegative : t -> t
+  (** [fp_isNegative f] test if it is a negative floating point *)
+
+  val fp_isPositive : t -> t
+  (** [fp_isPositive f] test if it is a positive floating point *)
+
+  val ieee_format_to_fp: int -> int -> t -> t
+  (** [ieee_format_to_fp e s bv] Convert a bitvector into a floating point using IEEE 754-2008 interchange format.
+      (reinterpret the bitvector into floating-point)
+  *)
+
+  val fp_to_fp: int -> int -> t -> t -> t
+  (** [fp_to_fp e s rm f] convert from one floating point format to another *)
+
+  val real_to_fp: int -> int -> t -> t -> t
+  (** [real_to_fp e s rm r] convert from a real *)
+
+  val sbv_to_fp: int -> int -> t -> t -> t
+  (** [sbv_to_fp e s rm bv] convert from a signed integer *)
+
+  val ubv_to_fp: int -> int -> t -> t -> t
+  (** [ubv_to_fp e s rm bv] convert from an unsigned integer *)
+
+  val to_ubv: int -> t -> t -> t
+  (** [to_ubv m rm f] convert to an unsigned integer (bitvector of size m) *)
+
+  val to_sbv: int -> t -> t -> t
+  (** [to_ubv m rm f] convert to a signed integer (bitvector of size m) *)
+
+  val to_real: t -> t
+  (** [to_real f] convert to a real *)
+
+  val type_for_to_fp: t -> [ `Real | ` Bitv | `Float | `Other ]
+  (** Used for `to_fp` parsing; give the broad type of the given term *)
```

Comment:
You're right about the name (though I actually don't like `type_of` very much as a name, for I don't know what reason, :p ). I guess it could be called `builtin_ty` instead ?

Also, we can note that there is already is a function `ty: term -> ty`, used by the arithmetic theory which also uses type-based dispatch:
https://github.com/Gbury/dolmen/blob/f2b68977003f58f82b1397dadcdf3e195ceab026/src/typecheck/arith.ml#L138-L145

The solution in arithmetic is clearly not optimal, and will not work here since bitvectors are a indexed family of types and not a single one, but once the `classify`/`builtin_ty` function is introduced, I'll adapt the arith code to use it as it will make its code better I think.

Reply:
Could the type (in `Dolmen.Expr`), be : ```classify: t -> [`Int | `Real | `Bitv of int | `Float of int * int | `Array of ty * ty] ```, and in the required interface for floats the type would be: ```classify: t -> [> `Real | `Bitv | `Float]``` ?
That way, we'd have a bit more generic way for theories to match on the type of a term, which I could re-use in the int/real theories. (I can do the change and push it to your branch if you prefer).

Reply:
Indeed, we'd probably need a catch_all case like ``` `Other```, I'll try to make that change tomorrow.

Reply:
Changes pushed. It ended up being a bit more ambitious than I thought at first, but the use of views on types seems much more clean now, ^^